### PR TITLE
Implement correlation protocol

### PR DESCRIPTION
**Pull request by mrproliu**
implement [correlation protocol](https://github.com/apache/skywalking/blob/master/docs/en/protocols/Skywalking-Cross-Process-Correlation-Headers-Protocol-v1.md), support transport and adding custom data.

### CODE REVIEW

**Comment by mrproliu** [comment]
After I add the document, it can start to review.

**Comment by mrproliu** [comment]
@wu-sheng @moonming I have finished the document and tests, please help to review.

File: README.md [hunk 1]
```diff
@@ -118,7 +120,7 @@ The following APIs are for developers or using this lib out of the Nginx case.
 
 ## Nginx APIs
 - **startTimer**, `require("client"):startBackendTimer("http://127.0.0.1:8080")`. Start the backend timer. This timer register the metadata and report traces to the backend.
-- **start**, `require("tracer"):start("upstream service")`. Begin the tracing before the upstream begin.
+- **start**, `require("tracer"):start("upstream service", correlation)`. Begin the tracing before the upstream begin. The custom data (table type) can be passed in the second parameter, and then you can correlate the data to the downstream service.
```
Comment (wu-sheng) [inline README.md]:
```suggestion
- **start**, `require("tracer"):start("upstream service", correlation)`. Begin the tracing before the upstream begin. The custom data (table type) can be injected as the second parameter, and then they will be propagated to the downstream service.
```

File: lib/skywalking/correlation_context.lua [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    for i, per_data in ipairs(data)
+    do
+        if #data > ELEMENT_MAX_NUMBER then
+            return context
+        end
+
+        local parts = Util.split(per_data, ':')
+        local key = decode_base64(parts[1])
+        local value = #parts > 1 and decode_base64(parts[2]) or ""
```
Comment (wu-sheng) [inline lib/skywalking/correlation_context.lua]: `#parts > 1` is a legal/expected case? Do we do this in Java agent implementation?

Comment (wu-sheng) [reply lib/skywalking/correlation_context.lua]: Reading the `#serialize`, I didn't see it coming.

File: lib/skywalking/correlation_context.lua [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+-- Return string to represent this correlation context
+function _M.serialize(context)
+    local encoded = ''
+    for name, value in pairs(context) do
+        if #encoded > 0 then
+            encoded = encoded .. ','
+        end
+
+        encoded = encoded .. encode_base64(name) .. ':' .. encode_base64(value)
```
Comment (wu-sheng) [inline lib/skywalking/correlation_context.lua:77]: Is `encode_base64` safe for empty string?

File: lib/skywalking/correlation_context.lua (hunk 2 above)
Comment (mrproliu) [reply lib/skywalking/correlation_context.lua]: In [java agent implementation](https://github.com/apache/skywalking/blob/569baf3236226b17c3b40e376e5681b5623abf38/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/CorrelationContext.java#L112), it still has an empty check.

File: lib/skywalking/correlation_context.lua (hunk 3 above)
Comment (mrproliu) [reply lib/skywalking/correlation_context.lua:77]: it's safe. I have tested in the correlation_context_test.lua file.

File: lib/skywalking/correlation_context.lua (hunk 2 above)
Comment (wu-sheng) [reply lib/skywalking/correlation_context.lua]: My question is more about, should we abandon this whole section, as the format is illegal.

Comment (mrproliu) [reply lib/skywalking/correlation_context.lua]: Ok, I think we should abandon it, let me fix it and java implementation.

Comment (wu-sheng) [reply lib/skywalking/correlation_context.lua]: Thanks.

**Review by wu-sheng — APPROVED** [review]
LGTM. @moonming recheck please.

File: lib/skywalking/correlation_context.lua [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+-- limit define
+local ELEMENT_MAX_NUMBER = 3;
+local VALUE_MAX_LENGTH = 128;
```
Comment (moonming) [inline lib/skywalking/correlation_context.lua]: code style: remove the `;`

**Review by moonming — APPROVED** [review]

**Pull request merged**